auto_submit/lib/requests/check_pull_request.dart

```diff
@@ -39,14 +44,61 @@ class CheckPullRequest extends RequestHandler {
       return responses;
     }
     final List<Future<Response>> futures = <Future<Response>>[];
+    //The repoPullRequestsMap stores the repo name and the set of PRs ready to merge to this repo
+    final Map<String, Set<PullRequest>> repoPullRequestsMap = <String, Set<PullRequest>>{};
     for (pub.ReceivedMessage message in receivedMessages) {
-      futures.add(_processMessage(message));
+      futures.add(_processMessage(message, repoPullRequestsMap));
     }
     responses.addAll(await Future.wait(futures));
+
+    await checkPullRequests(repoPullRequestsMap);
+    return responses;
+  }
+
+  /// Check and merge the pull requests to each repo this cycle.
+  ///
+  /// The number of pull requests to be merged to each repo will not exceed
+  /// the _kMergeCountPerRepo
+  Future<List<Response>> checkPullRequests(Map<String, Set<PullRequest>> repoPullRequestsMap) async {
+    final List<Response> responses = <Response>[];
+    for (String repoName in repoPullRequestsMap.keys) {
+      // Merge first _kMergeCountPerRepo counts of pull requests to each repo
+      for (int index = 0; index < repoPullRequestsMap[repoName]!.length; index++) {
+        final PullRequest pullRequest = repoPullRequestsMap[repoName]!.elementAt(index);
+        if (index < _kMergeCountPerRepo) {
+          final bool mergeResult = await _processMerge(pullRequest);
+          if (!mergeResult) {
+            responses
+                .add(Response.ok('Failed to merge the pull request ${pullRequest.number} to $repoName repository.'));
+          } else {
+            responses.add(
+                Response.ok('Successfully merged the pull request ${pullRequest.number} to $repoName repository.'));
+          }
+        } else {
+          await pubsub.publish('auto-submit-queue', repoPullRequestsMap[repoName]!.elementAt(index));
+          responses.add(Response.ok('Cannot merge the pull request ${pullRequest.number} to $repoName repository'
+              'this time because we already merged $_kMergeCountPerRepo pull requests to this repository.'));
```

Comment:
I would use a `map<string, string>` mapping like `pr number` to `status` to make the logic more clear. For example:
pr123: merged
pr456: unmerged
pr789: queued

Reply:
Gotcha Keyong, yes I think this can make the logic more clear.